src/main/php/PHPMD/Renderer/HTMLRenderer.php

```diff
@@ -2,155 +2,627 @@
+		$w->write("<h2 style='page-break-before: always'>Details</h2>");
+		$w->write("
+			<a
+				id='details-link'
+				class='info-lnk blck'
+				href='#'
+				onclick='toggle(\"details-link\"); toggle(\"details-wrapper\"); return false;'
+			>
+			Show details &#x25BC;
+		</a>");
+		$w->write("<div id='details-wrapper' class='hidden'>");
+
+		foreach ($violations as $violation) {
+
+			// This is going to be used as ID in HTML (deep anchoring).
+			$htmlId = "p-" . $index++;
+
+			// Get excerpt of the code from validated file.
+			$excerptHtml = null;
+			$excerpt = self::getLineExcerpt(
+				$violation->getFileName(),
+				$violation->getBeginLine(),
+				2
+			);
+
+			foreach ($excerpt as $line => $code) {
+				$class = $line === $violation->getBeginLine() ? " class='hlt'" : null;
+				$codeHtml = htmlspecialchars($code);
+				$excerptHtml .= "<li{$class}><div class='no'>{$line}</div><div class='cd'>{$codeHtml}</div></li>";
+			}
+
+			$descHtml = self::colorize(htmlentities($violation->getDescription()));
+			$filePath = $violation->getFileName();
+			$fileHtml = "<a href='file://$filePath' target='_blank'>" . self::highlightFile($filePath) . "</a>";
+
+			// Create an external link to rule's help, if there's any provided.
+			$linkHtml = null;
+			if ($url = $violation->getRule()->getExternalInfoUrl()) {
+				$linkHtml = "<a class='info-lnk' href='{$url}' target='_blank'>(help)</a>";
+			}
+
+			// HTML snippet handling the toggle to display the file's code.
+			$showCodeAnchor = "
+				<a class='info-lnk blck' href='#' onclick='toggle(\"$htmlId-code\"); return false;'>
+					Show code &#x25BC;
+				</a>";
+
+			$prio = self::$priority_titles[$violation->getRule()->getPriority()];
+			$html = "
+				<section class='prb' id='$htmlId'>
+					<header>
+						<h3>
+							<a href='#$htmlId' class='indx'>#{$index}</a>
+							{$descHtml} {$linkHtml} <span class='prio'>{$prio}</span>
+						</h3>
+					</header>
+					<div class='sub-info'><b>File:</b> {$fileHtml} {$showCodeAnchor}</div>
+					<ul class='code hidden' id='$htmlId-code'>%s</ul>
+				</section>";
+
+			// Remove unnecessary tab/space characters at the line beginnings.
+			$html = self::reduceWhitespace($html);
+			$w->write(sprintf($html, $excerptHtml));
+
+		}
+
+	}
+
+	/**
+	 * This method will be called the engine has finished the report processing
+	 * for all registered renderers.
+	 *
+	 * @return void
+	 */
+	public function end()
+	{
+		$writer = $this->getWriter();
+		$writer->write('</div></body></html>');
+	}
+
+	/**
+	 * Return array of lines from a specified file:line, optionally with extra lines around
+	 * for additional cognitive context.
+	 *
+	 * @return array
+	 */
+	protected static function getLineExcerpt($file, $lineNumber, $extra = 0)
+	{
+
+		if (!is_readable($file)) {
+			return [];
+		}
+
+		$file = new \SplFileObject($file);
+
+		// We have to subtract 1 to extract correct lines via SplFileObject.
+		$line = $lineNumber - 1 - $extra;
```

Comment:
I think this has to be changed to 
```suggestion
		$line = max($lineNumber - 1 - $extra, 0);
```
That way if there's an error on the first line we don't get an exception with seek -1